Extracted annotators to make them configurable

netvl opened the pull request:
This fixes #64 

Default behavior is not changed, so this PR is backwards-compatible. In order to override the annotation strategy an `Annotator` instance should be brought into scope:

```scala
package example

import upickle._
import upickle.annotator.key._

sealed trait Root
case class A(value: Int) extends Root

write(A(10))  // {"$variant": "example.A", "value": 10}
```

It is also possible to override the discriminator key:
```scala
package example

import upickle._

implicit val customKeyAnnotator = annotator.key.keyAnnotator("$custom$")

sealed trait Root
case class A(value: Int) extends Root

write(A(10))  // {"$custom$": "example.A", "value": 10}
```

If one wants to override the default annotator, the corresponding implicit has to be present in the scope where `Reader`s/`Writer`s are defined.

lihaoyi commented:
- I don't like default implicits in this case; too easy to accidentally leave out an import and have two parts of your program not-talking to each other. This was a problem earlier with the macro default RW accidentally being used to RW standard-library stuff for which there's already a RW in Implicits, before I made them imported by default, and it's probably still quite dangerous.I don't mind forcing people to do an `import upickle.config.default` at the top of each file if they don't customize it
- I'd prefer making your variant-encoding (using `$variant`) the `upickle.config.default`; we can leave the current behavior as `upickle.config.old`. This will involve changing most of the tests, but that should be straightforward.
- @tindzk this would be of interest to you in https://github.com/lihaoyi/upickle/issues/55. Do you want to try out @netvl's branch? He's got a pretty clever way of modifying how the case class serialization is done, and you could easily modify it to control the serialization as-a-whole and e.g. get your case-class-as-a-tuple behavior

netvl commented:
I agree, explicit configuration decisions are better; I just wanted to keep the backwards compatibility. If it is not a problem, then I'll remote the default.

> import upickle.config.default

Do you want some kind of general configuration? I'm asking this because I'm currently working on the customization of case class fields serialization mentioned in #43 (that `Filter` stuff). I don't have a clean design yet, but I guess it should be integrated with common configuration if there will be one.

lihaoyi commented:
Backwards compatibility would be nice, but as a pre-`1.0` library, I think we can make a few changes without people getting too upset ^_^

Yeah a general configuration mechanism would be ideal. @tindzk would like one as well. How general is open for debate: maybe someday we'll be serializing to protobuffers and other things, who knows! For now just customizing case-classes should be sufficient I guess

netvl commented:
(BTW, serialization to something else than `Js` tree, e.g. some foreign BSON ADT, would be a very nice thing to have in our project as well)

Ok, I guess for now I indeed can make `upickle.config.default` and `upickle.config.old` objects which for now will only contain annotators.

tindzk commented:
I'm glad you did something similar to what I tried for #59 (strangely, I couldn't get it working with ``implicitly``). Indeed, we should think of a general configuration method. @lihaoyi suggested to have a look at his branch https://github.com/lihaoyi/upickle/tree/cake. Once I'm back from my trip I will check whether the Cake approach could lead to a cleaner solution than implicits.

netvl commented:
@tindzk, I think it works because I don't pass the configuration object through macros - I just generate pickler code with `implicitly[Annotator]` which is resolved after the macro expansion, not during it.

I still need to wrap my head around the cake approach (I know what cake pattern is, but its usage in this context is still vague to me), but anyway implicits are lightweight and don't need refactoring. And when properly done they still allow cake-like composition of configuration.

netvl commented:
@lihaoyi, I've refactored annotators classes arrangement and added the configuration objects with implicits. Now the desired configuration has to be imported explicitly.

I think I also have to update the readme but I'll do it later in a separate commit, probably in this pull request as well.

netvl commented:
@lihaoyi, hmm, I'm now not sure if the readme has to be updated now. It holds instructions for 0.2.5 which is the current stable version. I think it should be updated before the new release.

Other than that, this PR is ready for merge.

tindzk commented on js/shared/main/scala/upickle/Annotator.scala:
```diff
@@ -0,0 +1,43 @@
+
+object Annotator {
+  // It drops down to array annotations if the serialized object is not represented as a `Js.Obj`
+  def keyAnnotator(key: String): Annotator = new Annotator {
```
Would it be possible to use a ``case class`` here?

netvl replied:
I'm not sure I understand what you mean :)

tindzk replied:
Ok, let me rephrase: Is there a specific reason why you didn't write: ``case class KeyAnnotator(key: String) extends Annotator { ... }``?

netvl replied:
Thanks for the clarification.

Well, case classes are semantically intended to be a data transfer objects - they are needed to keep and pass any data around. Annotator is not a data object, and `key` is just a parameter for its functionality. Hence it is not a case class.

netvl commented:
BTW, just for the information, using annotators now looks like this:
```scala
package example

import upickle._
import upickle.config.default._

sealed trait Root
case class A(value: Int) extends Root

write(A(10))  // {"$variant": "example.A", "value": 10}
```

With a custom key:
```scala
package example

import upickle._

object customConfig extends upickle.config.DefaultConfig {
  override implicit val annotator = Annotator.keyAnnotator("$custom$")
}
import customConfig._

sealed trait Root
case class A(value: Int) extends Root

write(A(10))  // {"$custom$": "example.A", "value": 10}
```

It allows writing custom configuration easily and leaves room for more configuration options.

netvl commented:
@lihaoyi, I'm very sorry to bother you, but do you plan to merge this? We would like to use uPickle in our project, but without this change we can't :(

lihaoyi commented:
@netvl yeah I do, but unfortunately, I have work/other things going on and haven't had time to review =/ I'll get to it as soon as I can. Perhaps publishing your own fork for the time being may be a reasonable workaround.

The hardships of free-time development...

netvl commented:
Yeah, that's what I did for now. Anyway, thank you very much for the library :)

knshiro commented:
@lihaoyi 
Thank you very much for your library
I would love to use this in [sbt-packer](https://github.com/en-japan/sbt-packer), is there any chance for a merge?

lihaoyi commented:
Rendered redundant with http://lihaoyi.github.io/upickle-pprint/upickle/#CustomConfiguration, not exactly the same but pretty close, including trivial support for all the examples provided here and more, e.g. the very-very-old case-classes-as-tuple serialization! @knshiro @netvl @tindzk FYI
